### ISSUE TITLE
[BUG] RuntimeError: Expected attn_mask dtype to be bool.

### ISSUE THREAD

**Issue by alanzty** [mlfoundations/open_clip#484]
https://github.com/mlfoundations/open_clip/blob/37b729bc69068daa7e860fb7dbcf1ef1d03a4185/src/open_clip/transformer.py#L580

New pytorch versions such as torch1.13.1 or pytorch2 has deprecate attn_mask value as float. Please address this issue. 
Thanks.

**Comment by rwightman** [comment]
@alanzty sorry where / when have float attn_mask been deprecated? this is news to me... I'm only aware of byte masks being deprecated.

**Comment by rwightman** [comment]
https://pytorch.org/docs/stable/generated/torch.nn.MultiheadAttention.html#torch.nn.MultiheadAttention.forward  .... if you're encountering a specific runtime error, need full call stack, environment details, use case, etc...

**Comment by alanzty** [comment]
@rwightman , Sorry Float is not deprecated. However, if you use CPU instead of CUDA with Pytorch 2. This will not work. That is because pytorch convert query to torch.bfloat16, which is not the same dtype as the float32 mask.
Pytorch2 uses a function called scaled_dot_product_attention. This function cannot handle mismatched dtype between query and mask. 

I am using cuda 11.7, torch==2.0.0, open_clip=2.16.1

**RuntimeError: Expected attn_mask dtype to be bool or to match query dtype, but got attn_mask.dtype: float and  query.dtype: c10::BFloat16 instead.**

**Comment by alanzty** [comment]
Here is the minimal example of the bug. 
import torch
# torch==2.0.0, cuda=11.7
b = 1
n = 64
d = 256

query = torch.randn(n, b, d)
key = torch.randn(n, b, d)
value = torch.randn(n, b, d)
attn_mask = torch.zeros(n, n)

attention = torch.nn.MultiheadAttention(d, 8)
with torch.no_grad(), torch.autocast("cpu"):
    output = attention(query, key, value, attn_mask=attn_mask, need_weights=False)

**Comment by rwightman** [comment]
@alanzty hmm, does it work without CPU autocast? either way seems like it might be a Pytorch bug not OpenCLIP, they've had numerous dtype issues with the masks in specific cases in the past. 

Switching to bool is definitely not a suitable answer, there are cases where float is needed, ie mixing bias into an additive mask...

**Comment by Amiralioghli** [comment]
I'm working on time series forecasting using Transformers in PyTorch. I encountered the following error. Can anyone help me? I'd really appreciate it!
```
forward
   attn_output, attn_output_weights = F.multi_head_attention_forward(

  File C:\ProgramData\anaconda3\Lib\site-packages\torch\nn\functional.py:5440 in multi_head_attention_forward
    attn_output = scaled_dot_product_attention(q, k, v, attn_mask, dropout_p, is_causal)

RuntimeError: Expected attn_mask dtype to be bool or to match query dtype, but got attn_mask.dtype: float and  query.dtype: double instead.
```

**Comment by abenicho** [comment]
if anyone came here for a pure `torch` error
@alanzty example returns the same error with `torch==2.1.2` and is solved by bumping to `torch==2.4.1`
I don't think it's connected to open_clip
